src/cogent3/core/new_alignment.py

```diff
@@ -4380,24 +4380,26 @@ def no_degenerates(self, motif_length: int = 1, allow_gap: bool = False):
             character (latter is default, as most evolutionary modelling treats
             gaps as N).
         """
-        try:
-            chars = list(self.moltype.alphabet)
-        except AttributeError as e:
-            msg = (
-                "Invalid MolType (no degenerate characters), "
+        if self.moltype.degen_alphabet is None:
+            raise new_moltype.MolTypeError(
+                f"Invalid MolType={self.moltype.label} (no degenerate characters), "
                 "create the alignment using DNA, RNA or PROTEIN"
             )
-            raise ValueError(msg) from e
+
+        chars = len(self.moltype)
 
         array_pos = self.array_positions
-        cutoff = len(chars) + 1 if allow_gap else len(chars)
+        # by design, char alphabets are organised such that the canonical
+        # characters always occur first, followed by gap, then ambiguity
+        # characters. so we can define a cutoff as follows:
+        cutoff = chars + 1 if allow_gap else chars
         indices = (array_pos < cutoff).all(axis=1)
 
         if motif_length > 1:
             num_motif = len(self) // motif_length
 
-            if len(self) % motif_length != 0:
-                indices = indices[: num_motif * motif_length]
+            if remainder := len(self) % motif_length:
+                indices = indices[:-remainder]
```

Comment:
you can probably delete `num_pos` and just slice array_pos the same as indices, i.e. `array_pos = array_pos[:-remainder]`

Reply:
I'll make the change and then merge